### PR TITLE
Implement a new preferences system inspired by the original one from …

### PR DESCRIPTION
**Pull request by nexus6-haiku**
…R5/Dano era

### CODE REVIEW

File: source/Preferences.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+#include "Preferences.h"
+
+
+Preferences::Preferences(BMessenger target)
+	: fTarget(target)
+{
+}
+
```
Comment (korli) [inline source/Preferences.cpp:13]:
```suggestion


```

File: source/Preferences.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+Preferences::~Preferences()
+{
+}
+
```
Comment (korli) [inline source/Preferences.cpp:17]:
```suggestion


```

File: source/Preferences.h [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+template <class T>
+void
+Preferences::SetProperty(BString property, T value)
+{
+	BMessage message(B_SET_PROPERTY);
+	message.AddSpecifier(property);
+	if constexpr (std::is_same<T, int32>::value) {
+		message.AddInt32("data", value);
+    } else if constexpr (std::is_same<T, bool>::value) {
```
Comment (korli) [inline source/Preferences.h:49]:
```suggestion
	} else if constexpr (std::is_same<T, bool>::value) {
```

File: source/Preferences.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+template <class T>
+status_t
+Preferences::SetTabProperty(int32 tabIndex, BString property, T value)
+{
+	BMessage message(B_SET_PROPERTY);
+	message.AddSpecifier(property);
+	message.AddSpecifier("tab", tabIndex);
+	if constexpr (std::is_same<T, int32>::value) {
+		message.AddInt32("data", value);
+    } else if constexpr (std::is_same<T, BString>::value) {
```
Comment (korli) [inline source/Preferences.h:151]:
```suggestion
	} else if constexpr (std::is_same<T, BString>::value) {
```

File: source/PreferencesWindow.h [hunk 5]
```diff
@@ -1,63 +1,83 @@
...
-#ifndef _PREFERENCES_WINDOW_H
-#define _PREFERENCES_WINDOW_H
 
+#pragma once
+
+#include "PreferencesWindow.h"
+#include "Preferences.h"
+#include "TabListView.h"
 
 #include <Box.h>
```
Comment (korli) [inline source/PreferencesWindow.h:12]:
```suggestion
```

File: source/PreferencesWindow.h [hunk 6]
```diff
@@ -1,63 +1,83 @@
...
 #include <Button.h>
 #include <CheckBox.h>
+#include <ColorControl.h>
```
Comment (korli) [inline source/PreferencesWindow.h:15]: Generally adding a class reference in the header is enough.
```suggestion
class BColorControl;
```

it's also the case for other members of the class.

File: source/TabListView.cpp [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+	fGlobalDropTargetIndicator(false),
+	fDeleteMessage(NULL),
+	fDragMessage(NULL)
+{
+}
+
+
+TabListView::~TabListView()
+{
+}
```
Comment (korli) [inline source/TabListView.cpp:44]:
```suggestion
delete fDeleteMessage;
delete fDragMessage;
}
```

Comment (korli) [reply source/TabListView.cpp:46]: Sorry, the tabs are missing.

Comment (nexus6-haiku) [reply source/TabListView.cpp:46]: fixed

**Comment by korli** [comment]
> @korli Are you happy with merging it or is there anything else you want me to fix?

Have you a screenshot?

**Comment by nexus6-haiku** [comment]
There is [this post](https://discuss.haiku-os.org/t/dockbert-improvements/15644/46). Is that ok?

**Comment by korli** [comment]
Looks good

**Pull request merged**